cmd/kubectl-gadget/utils/trace.go

```diff
@@ -290,32 +454,243 @@ RetryLoop:
 		// Don't print warnings if at least one node succeeded. This avoids showing
 		// warnings together with the actual output generated by other nodes.
 		if successNodeCount == 0 {
-			printTraceFeedback(contextLogger.Warningf, nodeWarnings)
+			printWarningFunction := func(format string, args ...interface{}) { fmt.Fprintf(os.Stderr, "[WARN] "+format+"\n", args...) }
+
+			printTraceFeedback(printWarningFunction, nodeWarnings)
 		}
 
+		printErrorFunction := func(format string, args ...interface{}) { fmt.Fprintf(os.Stderr, "[ERR]"+format+"\n", args...) }
+
 		// Print errors even if other nodes succeeded.
-		printTraceFeedback(contextLogger.Errorf, nodeErrors)
+		printTraceFeedback(printErrorFunction, nodeErrors)
 
 		if successNodeCount == 0 {
-			deleteTraces(contextLogger, traceRestClient, traceID)
-			contextLogger.Fatalf("Failed to run the gadget on all nodes: None of them succeeded")
+			return errors.New("Failed to run the gadget on all nodes: None of them succeeded")
 		}
+
 		break RetryLoop
 	}
 
+	client, err := k8sutil.NewClientsetFromConfigFlags(KubernetesConfigFlags)
+	if err != nil {
+		return fmt.Errorf("Error in creating setting up Kubernetes client: %w", err)
+	}
+
 	if customResultsDisplay == nil {
-		if outputMode != "Stream" {
-			panic(fmt.Errorf("OutputMode=%q needs a custom display function", outputMode))
+		// If we arrived here, it means we have at least one successful node, so it
+		// is safe to use trace pointer.
+		if trace.Spec.OutputMode != "Stream" {
+			return fmt.Errorf("OutputMode=%q needs a custom display function", trace.Spec.OutputMode)
 		}
-		genericStreamsDisplay(contextLogger, client, params, &results, transformLine)
+
+		params := &CommonFlags{
+			OutputMode: config.CommonFlags.OutputMode,
+			Node:       config.CommonFlags.Node,
+			Verbose:    config.CommonFlags.Verbose,
+		}
+
+		genericStreamsDisplay(client, params, &traces, transformLine)
 	} else {
-		customResultsDisplay(contextLogger, nodes, &results)
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("Error listing nodes: %w", err)
+		}
+
+		customResultsDisplay(nodes, &traces)
+	}
+
+	return nil
+}
+
+// DeleteTrace deletes the traces for the given trace ID using RESTClient.
+func DeleteTrace(traceID string) error {
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return err
+	}
+
+	deleteTraces(traceRestClient, traceID)
+
+	return nil
+}
+
+// This type is used to create a string containing labels to filter out trace.
+type labelFilter struct {
+	GadgetName string
+	NodeName string
+	Namespace string
+	PodName string
+	ContainerName string
+	OutputMode string
+}
+
+// labelsFromFilter creates a string containing labels value from the given
+// labelFilter.
+func labelsFromFilter(filter labelFilter) string {
+	labels := ""
+	separator := ""
+
+	v := reflect.ValueOf(filter)
```

Comment:
I don't think using reflect for this is the best approach. What about using a `map[string]string` instead of `labelFilter`. You could just loop through all the elements of the map to create the label string. 
The caller could use something like:

```
filter := map[string]string{
		"GadgetName": config.GadgetName,
		"NodeName":   config.CommonFlags.Node,
...
```